pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -230,13 +231,117 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
+	}
 	go adc.reconciler.Run(stopCh)
 	go adc.desiredStateOfWorldPopulator.Run(stopCh)
 
 	<-stopCh
 	glog.Infof("Shutting down Attach Detach Controller")
 }
 
+func (adc *attachDetachController) populateActualStateOfWorld() error {
+	glog.V(5).Infof("Populating ActualStateOfworld")
+	nodes, err := adc.kubeClient.Core().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for i := range nodes.Items {
+		nodeName := types.NodeName(nodes.Items[i].Name)
+		for _, attachedVolume := range nodes.Items[i].Status.VolumesAttached {
+			uniqueName := attachedVolume.Name
+			glog.V(5).Infof("Adding to ASW: volume %s, node %s", uniqueName, nodeName)
+			err = adc.actualStateOfWorld.MarkVolumeAsAttached(uniqueName, nil, nodeName, attachedVolume.DevicePath)
+			if err != nil {
+				glog.Errorf("Error adding node to ActualStateOfWorld: %v", err)
+				continue
+			}
+			adc.processVolumesInUse(nodeName, nodes.Items[i].Status.VolumesInUse, true)
+		}
+	}
+	return nil
+}
+
+func (adc *attachDetachController) getNodeVolumeDevicePath(
+	volumeName v1.UniqueVolumeName, nodeName types.NodeName) (string, error) {
+	var devicePath string = ""
+	var found bool = false
+	node, err := adc.kubeClient.Core().Nodes().Get(string(nodeName), metav1.GetOptions{})
+	if err != nil {
+		return devicePath, err
+	}
+	for _, attachedVolume := range node.Status.VolumesAttached {
+		if volumeName == attachedVolume.Name {
+			devicePath = attachedVolume.DevicePath
+			found = true
+			break
+		}
+	}
+	if !found {
+		err = fmt.Errorf("Volume %s not found on node %s", volumeName, nodeName)
+	}
+
+	return devicePath, err
+}
+
+func (adc *attachDetachController) populateDesiredStateOfWorld() error {
+	glog.V(5).Infof("Populating DesiredStateOfworld")
+
+	nodes, err := adc.kubeClient.Core().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for i := range nodes.Items {
+		nodeName := types.NodeName(nodes.Items[i].Name)
+		adc.desiredStateOfWorld.AddNode(nodeName)
+	}
+
+	pods, err := adc.kubeClient.Core().Pods(v1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods.Items {
+		podToAdd := pod
+		adc.podAdd(&podToAdd)
+		for _, podVolume := range podToAdd.Spec.Volumes {
+			// The volume specs present in the ActualStateOfWorld are nil, let's replace those
+			// with the correct ones found on pods. The present in the ASW with no corresponding
+			// pod will be detached and the spec is irrelevant.
+			volumeSpec, err := adc.createVolumeSpec(podVolume, podToAdd.Namespace)
+			if err != nil {
+				glog.V(10).Infof(
+					"Error creating spec for volume %q, pod %q/%q: %v",
+					podVolume.Name,
+					podToAdd.Namespace,
+					podToAdd.Name,
+					err)
+				continue
+			}
+			nodeName := types.NodeName(podToAdd.Spec.NodeName)
+			volumeName := v1.UniqueVolumeName(podVolume.Name)
+			if adc.actualStateOfWorld.VolumeNodeExists(volumeName, nodeName) {
```

Comment:
@tsmetana ping on this. is `volumeName` being passed to `VolumeNodeExists` function accurate?